example/FormatResp.py

```diff
@@ -8,6 +8,46 @@
 
 import prettytable
 
+from nebula3.data.DataObject import ValueWrapper
+
+
+def cast(val: ValueWrapper):
+    if val.is_empty():
+        return '__EMPTY__'
+    elif val.is_null():
+        return '__NULL__'
+    elif val.is_bool():
+        return val.as_bool()
+    elif val.is_int():
+        return val.as_int()
+    elif val.is_double():
+        return val.as_double()
+    elif val.is_string():
+        return val.as_string()
+    elif val.is_time():
+        return val.as_time()
+    elif val.is_date():
+        return val.as_date()
+    elif val.is_datetime():
+        return val.as_datetime()
+    elif val.is_list():
+        return [cast(x) for x in val.as_list()]
+    elif val.is_set():
+        return {cast(x) for x in val.as_set()}
+    elif val.is_map():
+        return {k: cast(v) for k, v in val.as_map()}
+    elif val.is_vertex():
+        return val.as_node()
+    elif val.is_edge():
+        return val.as_relationship()
+    elif val.is_path():
+        return val.as_path()
+    elif val.is_geography():
+        return val.as_geography()
```

Comment:
```python3
if val. is_geography():
```
we need to check all the function above? call almost 15 functions...

and I think its something like `O(n)`

can we make it to a `dict` that is O(1)

Reply:
Something like
```python3
val_type_dict = {
    "is_edge": is_edge,
    "is_vertex": is_vertex,
  }
val_cast_func = val_type_dict.get(val.val_type)
if val_cast_func:
    return val_cast_func()
```

Reply:
we could do something like this?

```python
from nebula3.data.DataObject import ValueWrapper
from nebula3.common.ttypes import Value


as_map = {
    Value.NVAL: "as_null",
    Value.__EMPTY__: "as_empty",
    Value.BVAL: "as_bool",
    Value.IVAL: "as_int",
    Value.FVAL: "as_double",
    Value.SVAL: "as_string",
    Value.LVAL: "as_list",
    Value.UVAL: "as_set",
    Value.MVAL: "as_map",
    Value.TVAL: "as_time",
    Value.DVAL: "as_date",
    Value.DTVAL: "as_datetime",
    Value.VVAL: "as_vertex",
    Value.EVAL: "as_edge",
    Value.PVAL: "as_path",
    Value.GGVAL: "as_geography",
    Value.DUVAL: "as_duration"}


def cast_with_dict(val: ValueWrapper):
    """
    implement cast() with a dict()
    """
    _type = val._value.getType()
    return getattr(val, as_map[_type])()
```

And it looks indeed optimal? As each condition is a function call, the complier is indeed not smart enough to underlying map it.

```python
In [15]: dis.dis(cast)
  5           0 LOAD_FAST                0 (val)
              2 LOAD_METHOD              0 (is_empty)
              4 CALL_METHOD              0
              6 POP_JUMP_IF_FALSE       12

  6           8 LOAD_CONST               1 ('__EMPTY__')
             10 RETURN_VALUE

  7     >>   12 LOAD_FAST                0 (val)
             14 LOAD_METHOD              1 (is_null)
             16 CALL_METHOD              0
             18 POP_JUMP_IF_FALSE       24

  8          20 LOAD_CONST               2 ('__NULL__')
             22 RETURN_VALUE

  9     >>   24 LOAD_FAST                0 (val)
             26 LOAD_METHOD              2 (is_bool)
             28 CALL_METHOD              0
             30 POP_JUMP_IF_FALSE       40

 10          32 LOAD_FAST                0 (val)
             34 LOAD_METHOD              3 (as_bool)
             36 CALL_METHOD              0
             38 RETURN_VALUE

 11     >>   40 LOAD_FAST                0 (val)
             42 LOAD_METHOD              4 (is_int)
             44 CALL_METHOD              0
             46 POP_JUMP_IF_FALSE       56

 12          48 LOAD_FAST                0 (val)
             50 LOAD_METHOD              5 (as_int)
             52 CALL_METHOD              0
             54 RETURN_VALUE

 13     >>   56 LOAD_FAST                0 (val)
             58 LOAD_METHOD              6 (is_double)
             60 CALL_METHOD              0
             62 POP_JUMP_IF_FALSE       72

 14          64 LOAD_FAST                0 (val)
             66 LOAD_METHOD              7 (as_double)
             68 CALL_METHOD              0
             70 RETURN_VALUE

 15     >>   72 LOAD_FAST                0 (val)
             74 LOAD_METHOD              8 (is_string)
             76 CALL_METHOD              0
             78 POP_JUMP_IF_FALSE       88

 16          80 LOAD_FAST                0 (val)
             82 LOAD_METHOD              9 (as_string)
             84 CALL_METHOD              0
             86 RETURN_VALUE

 17     >>   88 LOAD_FAST                0 (val)
             90 LOAD_METHOD             10 (is_time)
             92 CALL_METHOD              0
             94 POP_JUMP_IF_FALSE      104

 18          96 LOAD_FAST                0 (val)
             98 LOAD_METHOD             11 (as_time)
            100 CALL_METHOD              0
            102 RETURN_VALUE

 19     >>  104 LOAD_FAST                0 (val)
            106 LOAD_METHOD             12 (is_date)
            108 CALL_METHOD              0
            110 POP_JUMP_IF_FALSE      120

 20         112 LOAD_FAST                0 (val)
            114 LOAD_METHOD             13 (as_date)
            116 CALL_METHOD              0
            118 RETURN_VALUE

 21     >>  120 LOAD_FAST                0 (val)
            122 LOAD_METHOD             14 (is_datetime)
            124 CALL_METHOD              0
            126 POP_JUMP_IF_FALSE      136

 22         128 LOAD_FAST                0 (val)
            130 LOAD_METHOD             15 (as_datetime)
            132 CALL_METHOD              0
            134 RETURN_VALUE

 23     >>  136 LOAD_FAST                0 (val)
            138 LOAD_METHOD             16 (is_list)
            140 CALL_METHOD              0
            142 POP_JUMP_IF_FALSE      162

 24         144 LOAD_CONST               3 (<code object <listcomp> at 0x103040f50, file "<ipython-input-14-7f0c6a799d38>", line 24>)
            146 LOAD_CONST               4 ('cast.<locals>.<listcomp>')
            148 MAKE_FUNCTION            0
            150 LOAD_FAST                0 (val)
            152 LOAD_METHOD             17 (as_list)
            154 CALL_METHOD              0
            156 GET_ITER
            158 CALL_FUNCTION            1
            160 RETURN_VALUE

 25     >>  162 LOAD_FAST                0 (val)
            164 LOAD_METHOD             18 (is_set)
            166 CALL_METHOD              0
            168 POP_JUMP_IF_FALSE      188

 26         170 LOAD_CONST               5 (<code object <setcomp> at 0x1046d6030, file "<ipython-input-14-7f0c6a799d38>", line 26>)
            172 LOAD_CONST               6 ('cast.<locals>.<setcomp>')
            174 MAKE_FUNCTION            0
            176 LOAD_FAST                0 (val)
            178 LOAD_METHOD             19 (as_set)
            180 CALL_METHOD              0
            182 GET_ITER
            184 CALL_FUNCTION            1
            186 RETURN_VALUE

 27     >>  188 LOAD_FAST                0 (val)
            190 LOAD_METHOD             20 (is_map)
            192 CALL_METHOD              0
            194 POP_JUMP_IF_FALSE      214

 28         196 LOAD_CONST               7 (<code object <dictcomp> at 0x102ca4710, file "<ipython-input-14-7f0c6a799d38>", line 28>)
            198 LOAD_CONST               8 ('cast.<locals>.<dictcomp>')
            200 MAKE_FUNCTION            0
            202 LOAD_FAST                0 (val)
            204 LOAD_METHOD             21 (as_map)
            206 CALL_METHOD              0
            208 GET_ITER
            210 CALL_FUNCTION            1
            212 RETURN_VALUE

 29     >>  214 LOAD_FAST                0 (val)
            216 LOAD_METHOD             22 (is_vertex)
            218 CALL_METHOD              0
            220 EXTENDED_ARG             0
            222 POP_JUMP_IF_FALSE      232

 30         224 LOAD_FAST                0 (val)
            226 LOAD_METHOD             23 (as_node)
            228 CALL_METHOD              0
            230 RETURN_VALUE

 31     >>  232 LOAD_FAST                0 (val)
            234 LOAD_METHOD             24 (is_edge)
            236 CALL_METHOD              0
            238 EXTENDED_ARG             0
            240 POP_JUMP_IF_FALSE      250

 32         242 LOAD_FAST                0 (val)
            244 LOAD_METHOD             25 (as_relationship)
            246 CALL_METHOD              0
            248 RETURN_VALUE

 33     >>  250 LOAD_FAST                0 (val)
            252 LOAD_METHOD             26 (is_path)
            254 CALL_METHOD              0
            256 EXTENDED_ARG             1
            258 POP_JUMP_IF_FALSE      268

 34         260 LOAD_FAST                0 (val)
            262 LOAD_METHOD             27 (as_path)
            264 CALL_METHOD              0
            266 RETURN_VALUE

 35     >>  268 LOAD_FAST                0 (val)
            270 LOAD_METHOD             28 (is_geography)
            272 CALL_METHOD              0
            274 EXTENDED_ARG             1
            276 POP_JUMP_IF_FALSE      286

 36         278 LOAD_FAST                0 (val)
            280 LOAD_METHOD             29 (as_geography)
            282 CALL_METHOD              0
            284 RETURN_VALUE
        >>  286 LOAD_CONST               0 (None)
            288 RETURN_VALUE

Disassembly of <code object <listcomp> at 0x103040f50, file "<ipython-input-14-7f0c6a799d38>", line 24>:
 24           0 BUILD_LIST               0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                12 (to 18)
              6 STORE_FAST               1 (x)
              8 LOAD_GLOBAL              0 (cast)
             10 LOAD_FAST                1 (x)
             12 CALL_FUNCTION            1
             14 LIST_APPEND              2
             16 JUMP_ABSOLUTE            4
        >>   18 RETURN_VALUE

Disassembly of <code object <setcomp> at 0x1046d6030, file "<ipython-input-14-7f0c6a799d38>", line 26>:
 26           0 BUILD_SET                0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                12 (to 18)
              6 STORE_FAST               1 (x)
              8 LOAD_GLOBAL              0 (cast)
             10 LOAD_FAST                1 (x)
             12 CALL_FUNCTION            1
             14 SET_ADD                  2
             16 JUMP_ABSOLUTE            4
        >>   18 RETURN_VALUE

Disassembly of <code object <dictcomp> at 0x102ca4710, file "<ipython-input-14-7f0c6a799d38>", line 28>:
 28           0 BUILD_MAP                0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                18 (to 24)
              6 UNPACK_SEQUENCE          2
              8 STORE_FAST               1 (k)
             10 STORE_FAST               2 (v)
             12 LOAD_FAST                1 (k)
             14 LOAD_GLOBAL              0 (cast)
             16 LOAD_FAST                2 (v)
             18 CALL_FUNCTION            1
             20 MAP_ADD                  2
             22 JUMP_ABSOLUTE            4
        >>   24 RETURN_VALUE

In [16]: dis.dis(cast_with_dict)
 62           0 LOAD_FAST                0 (val)
              2 LOAD_ATTR                0 (_value)
              4 LOAD_METHOD              1 (getType)
              6 CALL_METHOD              0
              8 STORE_FAST               1 (_type)

 63          10 LOAD_GLOBAL              2 (getattr)
             12 LOAD_FAST                0 (val)
             14 LOAD_GLOBAL              3 (as_map)
             16 LOAD_FAST                1 (_type)
             18 BINARY_SUBSCR
             20 CALL_FUNCTION            2
             22 CALL_FUNCTION            0
             24 RETURN_VALUE
```

Reply:
> Something like
> 
> ```python
> val_type_dict = {
>     "is_edge": is_edge,
>     "is_vertex": is_vertex,
>     ......
> }
> val_cast_func = val_type_dict.get(val.val_type)
> if val_cast_func:
>     return val_cast_func()
> ```

Aha, saw your example after submitting mine.

Reply:
In my opinion, it would be great if `ValueWrapper` have this built-in method which can automatically casts types (automatically calls `as_xxx` ) to concrete python type (e.g., `int`, `Dict[str, float]`, `List[List[int]` ).

Reply:
> In my opinion, it would be great if `ValueWrapper` have this built-in method which can automatically casts types (automatically calls `as_xxx` ) to concrete python type (e.g., `int`, `Dict[str, float]`, `List[List[int]` ).

You are right, that's why you had to try to help with the casting example outside.

Are you interested in adding cast to `class ValueWrapper` in this PR or a separated one?

Reply:
I will create another PR for `cast` method, thank you for your help!

Reply:
Cool! You are always the most welcomed @alan890104 :)!
And thanks to @yihong0618 :)